src/main/scala/org/broadinstitute/dsde/workbench/sam/util/NewRelicShadowResultReporter.scala

```diff
@@ -0,0 +1,144 @@
+    ).toJson.compactPrint
+
+    val sizedLogMessage = if (logMessage.size > maxLogMessageSize) {
+      logMessage.take(maxLogMessageSize) + "<truncated>"
+    } else {
+      logMessage
+    }
+    sizedLogMessage
+  }
+
+  def makeString(value: Any): String = {
+    value match {
+      case col: Traversable[_] => col.mkString(",")
+      case _ => value.toString
+    }
+  }
+
+  /**
+    * Checks that both realResult and shadowResult match. Match is not strictly an equality check which makes this
+    * complicated. In the case of collections order it not important. Type T can be a hierarchy of case classes
+    * containing collections at any level or T can itself be a collection. This function unpacks collections and
+    * case classes and recursively calls resultsMatch on each element.
+    *
+    * Failures are also handled special. In the case of WorkbenchExceptionWithErrorReport, status codes must be the same,
+    * otherwise so long as both fail that is ok
+    *
+    * @param realResult
+    * @param shadowResult
+    * @tparam T
+    * @return
+    */
+  def resultsMatch[T](realResult: Either[Throwable, T], shadowResult: Either[Throwable, T]): MatchResult = {
+    (realResult, shadowResult) match {
+      case (Left(realThrowable: WorkbenchExceptionWithErrorReport), Left(shadowThrowable: WorkbenchExceptionWithErrorReport)) =>
+        // if they both failed with WorkbenchExceptionWithErrorReport make sure both have same status code
+        createMatchResult(realThrowable.errorReport.statusCode, shadowThrowable.errorReport.statusCode, "unequal error report status codes")
+
+      case (Left(realThrowable: WorkbenchExceptionWithErrorReport), Left(shadowThrowable)) =>
+        // the real failure has a status code but the shadow does not
+        MatchResult(false, Seq(s"real contained status code [${realThrowable.errorReport.statusCode}], shadow threw [${shadowThrowable.getClass.getName}] message [${shadowThrowable.getMessage}]"))
+
+      case (Left(_), Left(_)) =>
+        // both just plain old failures, that they both failed is enough
+        MatchResult(true, Seq.empty)
+
+      case (Right(realTraversable: Traversable[_]), Right(shadowTraversable: Traversable[_])) =>
+        // both traversables, we don't care about order, we care that each item in real matches an item in shadow, that
+        // there are the same number of occurrences, and both real and shadow have the same number of distinct items
+        val real = realTraversable.groupBy(x => x).map { case (key, values) => (key, values.size) }
+        val shadow = shadowTraversable.groupBy(x => x).map { case (key, values) => (key, values.size) }
+        val sizeMatch = createMatchResult(real.size, shadow.size, s"unequal distinct item count, [${real.mkString(",")}] vs [${shadow.mkString(",")}]")
+        val itemsMatch = real.keySet.map { realValue =>
+          shadow.keySet.find(shadowValue => resultsMatch(Right(realValue), Right(shadowValue)).matches) match {
```

Comment:
why recurse just to check the equality of the two occurrence counts? can't you just call `createMatchResult`

Reply:
plain equality does not take into account ignoring of order

Reply:
I tried to make this clear in the scaladoc, let me know if it needs more

Reply:
wait, so you're calling the items inside the keyset -- keys -- "value"?

Reply:
well, yes... they are the actual values...